File: backend-py/src/voxpilot/services/tools/glob_search.py
"""glob_search tool — find files matching a glob pattern."""

from pathlib import Path
from typing import Any, ClassVar

from voxpilot.services.tools.base import Tool

_SKIP_DIRS = {
    ".git",
    "__pycache__",
    "node_modules",
    ".venv",
    "venv",
    ".mypy_cache",
    ".pytest_cache",
    ".ruff_cache",
}

MAX_RESULTS = 500


class GlobSearchTool(Tool):
    """Find files matching a glob pattern."""

    name = "glob_search"
    description = (
        "Find files matching a glob pattern within the working directory. "
        "Returns matching file paths relative to the working directory. "
        "Use '**/' for recursive matching (e.g., '**/*.py' finds all Python files)."
    )
    parameters: ClassVar[dict[str, Any]] = {
        "type": "object",
        "properties": {
            "pattern": {
                "type": "string",
                "description": "Glob pattern to match files (e.g., '**/*.py', 'src/**/*.ts').",
            },
            "path": {
                "type": "string",
                "description": (
                    "Subdirectory to search within (relative to working directory). "
                    "Defaults to '.' (entire working directory)."
                ),
            },
        },
        "required": ["pattern"],
        "additionalProperties": False,
    }

    async def execute(self, arguments: dict[str, Any], work_dir: Path) -> str:
        """Find files matching a glob pattern."""
        pattern: str = arguments.get("pattern", "")
        if not pattern:
            return "Error: 'pattern' argument is required."

        raw_path: str = arguments.get("path", ".")
        resolved = self._resolve_path(raw_path, work_dir)
        if resolved is None:
            return f"Error: path '{raw_path}' is outside the working directory."
        if not resolved.exists():
            return f"Error: path '{raw_path}' does not exist."
        if not resolved.is_dir():
            return f"Error: '{raw_path}' is not a directory."

        try:
            raw_matches = sorted(resolved.glob(pattern))
        except ValueError as exc:
            return f"Error: invalid glob pattern '{pattern}': {exc}"

        # Filter out skip dirs and non-files
        results: list[str] = []
        work_resolved = work_dir.resolve()
        for match in raw_matches:
            if any(part in _SKIP_DIRS for part in match.parts):
                continue
            if not match.is_file():
                continue
            rel = match.relative_to(work_resolved)
            results.append(str(rel))
            if len(results) >= MAX_RESULTS:
                results.append(f"... (truncated at {MAX_RESULTS} results)")
                break

        if not results:
            return f"No files found matching pattern '{pattern}'."

        header = f"Found {len(results)} file(s) matching '{pattern}':\n"
        return header + "\n".join(results)

File: backend-ts/src/tools/glob-search.ts
import type { ChatCompletionTool, FunctionDefinition } from "openai/resources";
import { readdir, stat } from "node:fs/promises";
import type { Dirent } from "node:fs";
import { join, relative, resolve } from "node:path";
import type { Tool } from "./base";
import { resolvePath } from "./base";

const SKIP_DIRS = new Set([
  ".git",
  "__pycache__",
  "node_modules",
  ".venv",
  "venv",
  ".mypy_cache",
  ".pytest_cache",
  ".ruff_cache",
]);

const MAX_RESULTS = 500;

export class GlobSearchTool implements Tool {
  readonly requiresConfirmation = false;

  readonly definition: FunctionDefinition = {
    name: "glob_search",
    description:
      "Find files matching a glob pattern within the working directory. " +
      "Returns matching file paths relative to the working directory. " +
      "Use '**/' for recursive matching (e.g., '**/*.py' finds all Python files).",
    parameters: {
      type: "object",
      properties: {
        pattern: {
          type: "string",
          description:
            "Glob pattern to match files (e.g., '**/*.py', 'src/**/*.ts').",
        },
        path: {
          type: "string",
          description:
            "Subdirectory to search within (relative to working directory). " +
            "Defaults to '.' (entire working directory).",
        },
      },
      required: ["pattern"],
      additionalProperties: false,
    },
  };

  toOpenAiTool(): ChatCompletionTool {
    return { type: "function", function: this.definition };
  }

  async execute(
    args: Record<string, unknown>,
    workDir: string,
  ): Promise<string> {
    const pattern = (args.pattern as string | undefined) ?? "";
    if (!pattern) {
      return "Error: 'pattern' argument is required.";
    }

    const rawPath =
      typeof args.path === "string" && args.path !== "" ? args.path : ".";

    const resolved = await resolvePath(rawPath, workDir);
    if (resolved === null) {
      return `Error: path '${rawPath}' is outside the working directory.`;
    }

    let st: Awaited<ReturnType<typeof stat>>;
    try {
      st = await stat(resolved);
    } catch {
      return `Error: path '${rawPath}' does not exist.`;
    }

    if (!st.isDirectory()) {
      return `Error: '${rawPath}' is not a directory.`;
    }

    const absWorkDir = resolve(workDir);
    const allFiles = await this.collectFiles(resolved);
    const glob = new Bun.Glob(pattern);

    const results: string[] = [];
    for (const filePath of allFiles) {
      const rel = relative(resolved, filePath);
      if (glob.match(rel)) {
        const workRel = relative(absWorkDir, filePath);
        results.push(workRel);
        if (results.length >= MAX_RESULTS) {
          results.push(`... (truncated at ${MAX_RESULTS} results)`);
          break;
        }
      }
    }

    if (results.length === 0) {
      return `No files found matching pattern '${pattern}'.`;
    }

    const header = `Found ${results.length} file(s) matching '${pattern}':\n`;
    return header + results.join("\n");
  }

  private async collectFiles(root: string): Promise<string[]> {
    const files: string[] = [];
    await this.walkDir(root, files);
    files.sort();
    return files;
  }

  private async walkDir(dir: string, files: string[]): Promise<void> {
    let entries: Dirent[];
    try {
      entries = await readdir(dir, { withFileTypes: true, encoding: "utf-8" });
    } catch {
      return;
    }

    for (const entry of entries) {
      if (entry.isDirectory()) {
        if (SKIP_DIRS.has(entry.name)) continue;
        await this.walkDir(join(dir, entry.name), files);
      } else if (entry.isFile()) {
        files.push(join(dir, entry.name));
      }
    }
  }
}
